libr/include/r_util/r_event.h

```diff
@@ -0,0 +1,33 @@
+/* radare - LGPL - Copyright 2018 - pancake */
+
+#ifndef R_EVENT_H
+#define R_EVENT_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct r_event_t {
+	void *user;
+	RList *callbacks;
+} REvent;
+
+typedef enum {
+	R_EVENT_META_SET = 0,
+} REventType;
+
+typedef struct r_event_meta_t {
+	int type;
+	ut64 addr;
+	const char *string;
+} REventMeta;
+
+typedef void (*REventCallback)(REvent *ev, REventType type, void *data);
+
+R_API REvent *r_event_new(void *user);
+R_API void r_event_free(REvent *ev);
+R_API void r_event_hook(REvent *ev, REventCallback cb);
```

Comment:
this should have a `REventType type` as well. You want to register a listener for a particular type of event, not one that handles everything.

Reply:
my idea was to handle the event type inside the callback. what other ppl think about this? this way its more versatile and you can do everything in one hook, and not having to create 200 hooks to handle all the possible events and not be able to detect when. you are missing an event

Reply:
How many times do you think you'll have the same code that handles both R_EVENT_META_SET and, for example, R_EVENT_FUNCTION_ADD? I think these two pieces of code will be very different, so if you put all together in one hook it will just be a lot of "if (type == ...)". IMHO it's better to have X different functions that just do one thing, instead of one that does a lot of things. We already went through the `switch (X) { case A: case B: ...` path with commands, and it's now a mess, where more granular function would have been better. This is mostly the same IMHO.

Reply:
this reminds me to the discussion on merging callback handlers in the esil vm :P

Reply:
Yes, and you can still hook one callback to multiple event types and do a switch if it makes sense for the case, or hook separate callbacks for each type.